doc/developer/design/20230528_managed_clusters.md

```diff
@@ -0,0 +1,349 @@
+`r1`, `r2`, ..., `rN`, where `N` is the desired replication factor, all of
+the size specified by the `SIZE` option.
+
+For unmanaged clusters, `CREATE CLUSTER`'s behavior remains unchanged.
+
+### `ALTER CLUSTER`
+
+#### Changing size and replication factor
+
+We'll extend the `ALTER CLUSTER` statement to support altering the `SIZE` and
+`REPLICATION FACTOR` of a cluster using the usual syntax for `ALTER ... SET`
+commands:
+
+```sql
+-- Change the size of a cluster to `medium`.
+ALTER CLUSTER c SET (SIZE = 'medium');
+-- Change the replication factor of a cluster to 2.
+ALTER CLUSTER c SET (REPLICATION FACTOR = 2);
+-- Simultaneously change the size and replication factor of a cluster.
+ALTER CLUSTER c SET (SIZE = 'medium', REPLICATION FACTOR = 2);
+```
+
+When the command updates the cluster's size, Materialize will immediately drop
+all existing replicas of the cluster and recreate them with the same names with
+the new size.
+
+> **Warning**
+>
+> In the initial implementation, changing a cluster's size will result in
+> downtime while the new replicas rehydrate. This may be surprising to users.
+>
+> We hope to change the behavior in a future release so that no downtime is
+> incurred, but this will require [dynamic cluster scheduling].
+
+When the command increases the cluster's replication factor, Materialize will
+immediately create as many new replicas as necessary to meet the desired
+replication factor. It will name the replicas such that the resulting replica
+set adheres to the `r1`, `r2`, ..., `rN` naming scheme.
+
+When the command decreases the cluster's replication factor, Materialize will
+immediately drop as many replicas as necessary to meet the desired replication
+factor. It will choose replicas to drop so that the resulting replica set
+adheres to the `r1`, `r2`, ..., `rN` naming scheme.
+
+Users may only modify these options for managed clusters. Attempting to set the
+size or replication factor for an unmanaged cluster will result in an error like
+the following:
+
+```
+> CREATE CLUSTER c REPLICAS ();
+> ALTER CLUSTER c SET (SIZE = 'medium');
+ERROR: cannot set SIZE option for unmanaged cluster "c"
+HINT: Manually manage replicas using CREATE CLUSTER REPLICA and DROP CLUSTER REPLICA,
+or convert the cluster to a managed cluster by running ALTER CLUSTER c SET (MANAGED).
+```
+
+#### Converting a managed cluster to an unmanaged cluster
+
+`ALTER CLUSTER` will support enabling the `MANAGED` option for a cluster to
+convert an unmanaged cluster to a managed cluster:
+
+```sql
+ALTER CLUSTER c SET (MANAGED = true);
+ALTER CLUSTER c SET (MANAGED); -- Shorthand for the above
+```
+
+Conversion is possible if and only if all of the following conditions are met:
+
+  * All replicas of the cluster have the same size `S`.
+  * The replicas are named `r1`, `r2`, ..., `rN`, with no gaps in the numbering.
+    (Zero replicas trivially matches this pattern with `N=0`.)
```

Comment:
I was mostly trying to optimize for "what is easiest to explain to users" slash "what is easiest for users to understand", though I wouldn't say this design nails it, as you've identified!

If we drop the lexicographically greatest replica, that seems a bit harder to explain than "we drop the replica with the largest `N`" 

The reason I was thinking we should force users to rename replicas ahead of `ALTER ... SET (MANAGED)` is so that you could track how the replica names evolved. If `my_replica` is a semantically significant replica for, idk, monitoring via replica-targeted queries, forcing the user to rename it to e.g. `r2` is hopefully a reminder to the user that they need to update their replica-targeted queries for the new name. 

It occurs to me that there's a potential middle ground here: `ALTER ... SET (MANAGED)` will automatically rename your replicas to match the required pattern, and leaves entries in the audit logs about the rename. At least that way you have the breadcrumbs to figure out what your replicas were renamed to.

Reply:
Yeah, exactly! It's backwards compatible to lift the restriction in a future release of Materialize, if it winds up being a problem.